tiledb/sm/serialization/query.cc

```diff
@@ -2735,7 +2735,13 @@ Status global_write_state_from_capnp(
       }
     }
     if (frag_meta_reader.hasFragmentUri()) {
-      frag_meta->fragment_uri() = URI(frag_meta_reader.getFragmentUri().cStr());
+      // Reconstruct the fragment uri out of the received fragment name
+      auto write_version = query.array()->array_schema_latest().write_version();
+      auto frag_dir_uri =
+          query.array()->array_directory().get_fragments_dir(write_version);
```

Comment:
We'll have a problem with this one :( If you recall I have moved `fragment_metadata` serialization out of query into a separate file in order to use it in other places as well, like `fragment_info` serialization and in the future `array` seralization. (see https://github.com/TileDB-Inc/TileDB/pull/3530)
To do that I have removed the dependency on the query object, and I am just passing the `array_schema` (that is usually available) in the `fragment_metadata_from_capnp` function. So here I am really not sure how we can reconstruct this URI without access to `array_directory` ..

Reply:
I tried this instead and it seems to work in my case, you could try something similar as well (you need the `array_uri` ) :
```
      // Reconstruct the relative fragment URI to a full path URI
      auto frag_dir_uri = array_uri.add_trailing_slash();
      if (fragment_info->array_schema_latest()->write_version() >= 12) {
        frag_dir_uri = frag_dir_uri.join_path(constants::array_fragments_dir_name);
      }
      fragment_info->to_vacuum().emplace_back(frag_dir_uri.join_path(uri.cStr()));
    }

```

Reply:
This seems ok to me except one thing. The `write_version > 12` detail seems like something that should live in only one place, ArrayDirectoy. Do you think it'd be a good idea to write a static function in ArrayDirectory that returns the `frag_dir_uri` based on an URI arg? The current `ArrayDirectory::get_fragment_dir()` can just call that function, i.e. `generate_fragment_dir_uri(write_version, uri_);` and serialization can call the same like `ArrayDirectory::generate_fragment_dir_uri(write_version, array_uri);`.

Reply:
Sure good idea!